### ISSUE TITLE
Certificate management should be done generically across all resources

### ISSUE THREAD

**Issue by vito** [concourse/concourse#1027]
(I've preemptively labeled this with `anti/contributor-burden` and `anti/worker-state` as both are relevant to this issue as trade-offs, depending on the approach.)

## What challenge are you facing?

Today, each resource which talks to an external service (i.e. all resources) have some form of the following configuration:

* `insecure_skip_verify` / `ignore_ssl` / `skip_ssl_verification` / etc.
* `ca_cert` / `ca_certs`

This is a substantial burden to both resource authors and users. Every author must implement these, and ideally they'd all agree on a naming convention (we've already lost that battle). Every user must also then configure their certs across all resources, which is especially annoying if the format is inconsistent.

## A Modest Proposal

One proposal would be to just bind-mount `/etc/ssl/certs`, read-only, into each resource container. We actually tried this already, and it does work, with some gotchas:

* `/etc/ssl/certs` often contains symlinks to other paths outside of `/etc/ssl/certs`, requiring them to be bind-mounted, too.
* `/etc/ssl/certs` is platform-specific, although it seems that it's common for platforms that store them in another location to at least bind-mount `/etc/ssl/certs` to it (i.e. Fedora).
* It's possible that `/etc/ssl/certs` or, more likely, the symlinks contained therein to clobber some part of the resource rootfs.

Other caveats being that you then have to configure CA certs on your workers, which on face value goes against Concourse's "stateless workers" mantra, but arguably it's similar in spirit to proxy configuration, i.e. it's configuration the worker needs to reach the network, and so the worker is the source of truth. This is especially true if the worker is configured with a man-in-the-middle SSL proxy, for example - the cert for the SSL proxy has to be configured.

## Call for alternate proposals

We've already explored this particular proposal, so this issue is both to surface our goals and to foster discussion around alternative approaches.

**Comment by vito** [comment]
@julz Have you or @emalm thought about this (the original proposal) before as a potential Garden feature? Certificate propagation is a problem that both Concourse and Diego have had to tackle, and it's a lot of platform-specific orchestration which kind of leaks through the Garden abstraction level. This feels somewhat similar, although much more wart-covered, with propagating `/etc/hosts` and `/etc/resolv.conf` in.

**Comment by haydonryan** [comment]
I'm hitting this issue at a large enterprise security conscious customer.  We are using concourse to deploy PCF, in vSphere environment.  They use their internal CA certs to sign all internal facing resources including their bitbucket server.  Because there is no ability to add a CA cert, they're having to disable SSL, which violates their security conventions.  Most on prem customers use internally signed certs so this would be an important feature for the enterprise.

**Comment by billimek** [comment]
Same boat as @haydonryan here.

If I'm reading the issue correctly it would require managing the installation of that cert to the host running the concourse worker (and somehow resolving the rats-nest of symlinks?) in order to solve this using the proposed solution.

**Comment by DRuggeri** [comment]
Yeah, this is painful, @haydonryan. I'm also suffering from the same affliction.

**Comment by chendrix** [comment]
concourse/semver-resource#31 attempted to add `skip_ssl_verification` to the semver-resource

**Comment by bonzofenix** [comment]
@haydonryan @vito we are hitting this issue at Allstate. We are encrypting all our internal traffic at a firewall level with a custom CA that needs to be trusted by all resources. :(

**Comment by pn-santos** [comment]
We at Yoti also use a custom CA for our internal docker registry... it's a pain to need to include `ca_certs` in every pipeline and task definition in docker image resources.

**Comment by pivotal-topher-bullock** [comment]
Considering pushing support for container cert mounting down to Garden ; cloudfoundry/guardian#83

**Comment by DRuggeri** [comment]
@vito, thinking more about the challenges you shared, it seems like it would be sane to use a combination of what @pivotal-topher-bullock mentions as well as a configurable param. Bear with me...

For concourse-owned resources that are compiled and distributed along with concourse (think: BOSH deployed stuff or just the downloaded binary), it's much easier to bind-mount a directory or just push the CA cert file itself because those containers are under your control (no questions arise about where the CA trust path is inside the container). On the other hand, one resource in particular poses a challenge - the docker-image resource. In that case, concourse has no way of knowing what kind of distro is inside so using /etc/ssl/certs may not work at all.

With that in mind, maybe an argument common across all resources (interpreted by concourse rather than by the individual resources to avoid maintenance burden of custom resource developers) could exist that offers two options along the lines of "just bind /etc/ssl/certs to /etc/ssl/certs" or "bind /some/source/directory to /some/target/directory".

Other options a la throwing spaghetti at the wall and seeing what sticks:

- Specify a trust file of pem-encoded certs that is just copied into to the container's FS at a specific location. Custom resources and docker images then just need to be made aware of the location of the all-in-one trust file. For concourse-owned resources, this could just replace /etc/ssl/certs/ca-certificates.crt. In my experiences, this is usually the only file that matters. A quick sanity check shows that [Go's custom SSL implementation](https://golang.org/src/crypto/x509/root_linux.go) doesn't use the openssl hashed certs at all, so at the least this greatly cleans up house for the concourse-owned stuff.
- On startup of a concourse worker, do a directory walk resolving symlinks (to a deterministic depth - maybe 2 or 3) of /etc/ssl/certs and create a "flattened" copy of it in the working directory. Bind mount or copy that directory structure into the container. This shouldn't be terribly expensive... at current check there are 347 files at about 2Mb in size and a worker ought to have plenty of horsepower to handle that. An enhancement could be to preserve symlinks that point to neighboring files and only make copies of files that are outside of the directory

**Comment by calebwashburn** [comment]
This would be nice if a cert is on the worker cert trusted certs then those should be copied into the containers trusted certs.  If you manage concourse with bosh then you can use the trusted certs feature of bosh to add custom trusted certs to each VM and then when concourse workers create the container to run a given resource in include those certs in the appropriate location.

**Comment by mariash** [comment]
@DRuggeri thanks for your input. I like the idea of passing certificates as a file to some specified location and maybe providing a command to run to install them. Please see my suggestion here: https://github.com/cloudfoundry/guardian/issues/83#issuecomment-307523651

**Comment by vito** [comment]
@DRuggeri Agreed. I appreciate that you called out that resource authors shouldn't be burdened with having to deal with the certs being mounted in different places. That was one of the the trickiest parts of this whole ordeal. I'm not sure what you mean by your first proposal though (the option to mind `/etc/ssl/certs` or some other directory) - where would that option live such that resource authors could code against only a single mount? Or would it be provided as an argument?

Per your other recommendations, I prefer the latter the most. In our initial approach I'm pretty sure we started with just the all-in-one trust file, but I think we ran in to trouble with tools that don't support just having that file. Things like `curl`. I don't want to assume that all resources are written in Go. I think a one-time import, similar to how we import the base resource types into BaggageClaim, would work out pretty well.

(But yeah, if we can get this handled by Garden in some way, a lot of this discussion may go away.)

**Comment by DRuggeri** [comment]
Sorry - been AFK a bit on vacation and then too much time spent trying to catch up.

Just for clarity, my point about GO was that bind-mounting a single file for the things distributed with the concourse binary would work since those are written in GO and the concourse project has full control of how/where those resources look for certs. This can address the first part of your proposal where each of the concourse-provided resources does SSL trust a bit differently (or not at all). Agreed that it isn't portable enough for a general purpose solution.

To answer the question about minding a specific directory, it's more or less a suggestion for a worst-case-scenario where the author of the custom resource has to do some acrobatics no matter what. Having the ability for the docker-image resource to mount or copy the certs from the worker into a directory specified by the author makes it so they can choose how to handle them. Or, better yet, if the mounted directory has everything they need, it could just simply be used as-is (binding over /etc/ssl/certs, for example). Although this adds burden for the custom resource developer, it is a smaller burden than having to bake their certs into the container image, IMO.

The more I think about it, a combination of (optionally) specifying a directory and doing a "flattening" of the worker's trust files on startup seems to resonate the the most. When the directory to mount the flattened certs is not set, no action is taken... but if it is set, that's where concourse binds the directory RO. This combination of options has the benefit of being backwards compatible for existing container images that have already baked their CA certs into the right location, as well as providing a path forward for those who would like to have the worker's CA certs pushed into the container (to take advantage of BOSH-managed certs, as an example).

**Comment by pn-santos** [comment]
Not sure if it's in the scope of this issue or if a solution would also address this, but another problem I found associated with certs signed by private CAs is that, if ATC is using a cert signed by a private CA,  `fly execute` will fail if the task requires inputs/outputs because of the untrusted certificate 
```
executing build 906
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (60) SSL certificate problem: unable to get local issuer certificate
```

**Comment by DRuggeri** [comment]
It's been a while since hearing anything... Any thoughts?

**Comment by labrown** [comment]
Also work at large financial institution with serious security awareness.  We **need** this feature.

**Comment by vito** [comment]
We had a bit of a hiatus on our Operations stream but have prioritized this as a further discussion item with the Garden team.

Part of the reason we opened this issue was to ask for more feedback and proposals from the community. As mentioned above, we already explored one approach, and would like more collaboration or at least characterizations of different use cases before trying another.

Everyone, please add GitHub reactions to the original issue instead of leaving comments saying how important it is - we're well aware! 😄

**Comment by labrown** [comment]
I can echo haydonryan's comment of May 2nd pretty much exactly.  I'm in the same situation using concourse to deploy PCF in an environment with pervasive internal self-signed certificates.

I like calebwashburn's suggestion of June 2nd, using bosh to insert trusted certificates into the workers and then having all images pick up those certs automatically.  That would solve my issue.

**Comment by DRuggeri** [comment]
Thanks for checking in @vito. I understand your point about the "me too's".
I'd like to follow up and inquire: is there anything we from the community/as consumers can help? Are the use cases clear enough? Has enough been said about how trust is managed across the use cases (like: BOSH managed, just plain old ca-certs file, or as a separate file altogether)? Do you need testers or folks to bounce ideas off of?

I'm very eager to see this available so am happy to help however I can... IF I can

**Comment by topherbullock** [comment]
Here's the current plan for tackling this.
- Create a magical 'ssl-certs' volume on each worker on startup
- Hydrate the `ssl-certs` volume by either:
   1

1. 

. 
   2.

**Comment by topherbullock** [comment]
Here's the current plan for tackling this:
1. Create a magical `ssl-certs` volume on each worker on startup for certs to live in
1. Add certs to the the `ssl-certs` volume by either :
   - a) creating the volume using the import strategy, and importing from the workers' `/etc/ssl/certs` directory (resolving and copying over all symlinked files) 
   - b) streaming in a folder of certs configured on the ATC
1. Bind-mount the `ssl-certs` volume on the worker at `/etc/ssl/certs` when creating any resource containers

The options for hydrating the volume with certs - `2a` and `2b` - each have their own tradeoffs and impact on how this feature is used:
**2a** is probably the most 'Concourse'-y option in that it maintains the rule of workers being externally managed, telling the ATC their specific network configuration or hardware capabilities via tags. 
**2b** would allow an operator to ensure any worker which registers has the certs required, and would reduce the operations overhead, and also means workers don't have any state on them which can affect how resources behave on individual workers.

**Comment by labrown** [comment]
I'm still new to concourse, but I like option **2b** better.  "just works" is easier to maintain, if you ask me.

**Comment by jama-pivotal** [comment]
Going to make this separate issues to track the "work to be done™":
- Look into how we can create a volume on the workers when they start (import volume)
- Look into both scenarios for getting the certs on the workers
- Test that the resources work using the certs in certain scenarios (MITM SSL Proxy, Private Docker Registry, Minio, etc)

**Comment by vito** [comment]
🎊 The eagle has landed! Accepted via BOSH and also via binaries on Fedora.

This will be shipped in 3.9, on by default for the BOSH release and opt-in for the binaries (since we can't assume as much about how they're deployed/configured). We can polish this up a bit more and have it on by default in the binaries in the future.
